internal/kafka/internal/workers/kafka_mgrs/kafkas_mgr.go

```diff
@@ -374,3 +388,116 @@ func (k *KafkaManager) updateClusterStatusCapacityMaxMetric(c services.KafkaStre
+func (k *KafkaManager) updateExpiresAtBasedOnQuotaEntitlement(kafka *dbapi.KafkaRequest, isQuotaEntitlementActive bool) error {
+	// if quota entitlement is active, ensure expires_at is set to null
+	if isQuotaEntitlementActive && kafka.ExpiresAt.Valid {
+		logger.Logger.Infof("updating expiration date of kafka instance %q to NULL", kafka.ID)
+		return k.updateKafkaExpirationDate(kafka, nil)
+	}
+
+	// if quota entitlement is not valid and expires_at is not already set, set its value based on the current time and grace period allowance
+	// note that there is a temporary check for zero value of time.Time for cases where the kafka instance expires_at value
+	// hasn't been updated null yet after migration/rollout. This additional check can be removed once all Kafka instances
+	// has been successfully migrated to the new expires_at type.
+	if !isQuotaEntitlementActive && (!kafka.ExpiresAt.Valid || kafka.ExpiresAt.Time.IsZero()) {
+		billingModel, err := k.kafkaConfig.GetBillingModelByID(kafka.InstanceType, kafka.ActualKafkaBillingModel)
+		if err != nil {
+			return err
+		}
+
+		// set expires_at to now + grace period days
+		expiresAtTime := time.Now().Add(time.Duration(billingModel.GracePeriodDays*86400) * time.Second)
+		logger.Logger.Infof("quota entitlement for kafka instance %q is no longer active, updating expires_at to %q", kafka.ID, expiresAtTime.Format(time.RFC1123Z))
+		return k.updateKafkaExpirationDate(kafka, &expiresAtTime)
+	}
+
+	logger.Logger.Infof("no expires_at changes needed for kafka %q, skipping update", kafka.ID)
+	return nil
+}
+
+// Updates expires_at field based on the lifespan seconds configured per Kafka instance type/size
+func (k *KafkaManager) updateExpiresAtBasedOnLifespanSeconds(kafka *dbapi.KafkaRequest, instanceSize config.KafkaInstanceSize) error {
+	// note that there is a temporary check for zero value of time.Time in cases where the kafka instance expires_at value
+	// hasn't been updated to null yet after migration/rollout. This additional check can be removed once all Kafka instances
+	// has been successfully migrated to the new expires_at type
+	if !kafka.ExpiresAt.Valid || kafka.ExpiresAt.Time.IsZero() {
+		// set expires_at to created_at + lifespanSeconds
+		expiresAtTime := kafka.CreatedAt.Add(time.Duration(*instanceSize.LifespanSeconds) * time.Second)
+		logger.Logger.Infof("lifespanSeconds defined for Kafka %q, updating expires_at to %q as it is not yet set", kafka.ID, expiresAtTime.Format(time.RFC1123Z))
+		return k.updateKafkaExpirationDate(kafka, &expiresAtTime)
+	}
+
+	logger.Logger.Infof("lifespanSeconds defined for Kafka %q but expires_at is already set, skipping update", kafka.ID)
+	return nil
+}
+
+// updates the expires_at field for the given Kafka instance
+func (k *KafkaManager) updateKafkaExpirationDate(kafka *dbapi.KafkaRequest, expiresAtTime *time.Time) error {
+	var expiresAt sql.NullTime
+	if expiresAtTime != nil {
+		_ = expiresAt.Scan(*expiresAtTime)
```

Comment:
I think there is no need to use scan, and thus we can avoid the possibility of an error

The following would be valid:
expiresAt = sql.NullTime{Time: *expiresAtTime, Valid: true}

Reply:
cc @JameelB @ziccardi to double check what I said is accurate

Reply:
Agree

Reply:
yep that works 👍  updated it to use that instead.